diff --git a/extern/CMakeLists.txt b/extern/CMakeLists.txt
index 1c051bff6105501ef995af0851852137ae7be24c..f7bf9c4556bbdced4d44ebb3343c834c1f8291c3 100644
--- a/extern/CMakeLists.txt
+++ b/extern/CMakeLists.txt
@@ -271,7 +271,7 @@ if (NOT OPENMW_USE_SYSTEM_ICU)
         SOURCE_DIR fetched/icu
         CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env ${ICU_ENV}
             <SOURCE_DIR>/icu4c/source/configure --enable-static --disable-shared
-            --disable-tests --disable-samples --disable-icuio --disable-extras ${ICU_ADDITIONAL_OPTS}
+            --disable-tests --disable-samples --disable-icuio --disable-extras --disable-tools ${ICU_ADDITIONAL_OPTS}
         BUILD_COMMAND make
         INSTALL_COMMAND ""
     )
