diff --git a/apps/openmw/mwgui/loadingscreen.cpp b/apps/openmw/mwgui/loadingscreen.cpp
index fd7afa2d8ab3d20c9aa8c8cef746f1685e9fb33d..f84d9a71a5e0679ae83093cb71f086005203bed2 100644
--- a/apps/openmw/mwgui/loadingscreen.cpp
+++ b/apps/openmw/mwgui/loadingscreen.cpp
@@ -287,6 +287,7 @@ namespace MWGui
 
     void LoadingScreen::setupCopyFramebufferToTextureCallback()
     {
+        return;
         // Copy the current framebuffer onto a texture and display that texture as the background image
         // Note, we could also set the camera to disable clearing and have the background image transparent,
         // but then we get shaking effects on buffer swaps.
