diff --git a/apps/openmw/engine.cpp b/apps/openmw/engine.cpp
index cd32cb3b0d50257ce35349b46ebab5f298864717..da41ca37e403f664351f5a30338351aed2ce7198 100644
--- a/apps/openmw/engine.cpp
+++ b/apps/openmw/engine.cpp
@@ -518,6 +518,9 @@ void OMW::Engine::setSkipMenu (bool skipMenu, bool newGame)
     mNewGame = newGame;
 }
 
+// To share the viewer with Android interfaces
+osg::ref_ptr<osgViewer::Viewer> g_viewer;
+
 void OMW::Engine::createWindow(Settings::Manager& settings)
 {
     int screen = settings.getInt("screen", "Video");
@@ -643,6 +646,9 @@ void OMW::Engine::createWindow(Settings::Manager& settings)
     mViewer->realize();
 
     mViewer->getEventQueue()->getCurrentEventState()->setWindowRectangle(0, 0, graphicsWindow->getTraits()->width, graphicsWindow->getTraits()->height);
+
+    // To share the viewer with Android interfaces
+    g_viewer = mViewer;    
 }
 
 void OMW::Engine::setWindowIcon()
@@ -1076,6 +1082,8 @@ void OMW::Engine::go()
 
     luaWorker.join();
 
+    g_viewer.release();
+    
     // Save user settings
     settings.saveUser(settingspath);
     mLuaManager->savePermanentStorage(mCfgMgr.getUserConfigPath().string());
diff --git a/components/debug/debugging.cpp b/components/debug/debugging.cpp
index c84bb39d3d12ba38223e2c3471b32fd74101d62f..e84624e2c94036ae3b3e13f7398dac59810f87da 100644
--- a/components/debug/debugging.cpp
+++ b/components/debug/debugging.cpp
@@ -13,6 +13,8 @@
 #   include <windows.h>
 #endif
 
+#include <osgViewer/Viewer>
+
 namespace Debug
 {
 #ifdef _WIN32
@@ -172,6 +174,8 @@ void setupLogging(const std::string& logDir, const std::string& appName, std::io
 #endif
 }
 
+extern osg::ref_ptr<osgViewer::Viewer> g_viewer;
+
 int wrapApplication(int (*innerApplication)(int argc, char *argv[]), int argc, char *argv[],
                     const std::string& appName, bool autoSetupLogging)
 {
@@ -209,7 +213,8 @@ int wrapApplication(int (*innerApplication)(int argc, char *argv[]), int argc, c
         ret = innerApplication(argc, argv);
     }
     catch (const std::exception& e)
-    {
+    {    	
+        g_viewer.release();    	
 #if (defined(__APPLE__) || defined(__linux) || defined(__unix) || defined(__posix))
         if (!isatty(fileno(stdin)))
 #endif
