diff --git a/apps/openmw/engine.cpp b/apps/openmw/engine.cpp
index 3e7f151fadfe13ee11ffc533f9a6b0ec32fcdb09..8c15f1f223783ef2a1e3cb1fdc4e26d9a1b08049 100644
--- a/apps/openmw/engine.cpp
+++ b/apps/openmw/engine.cpp
@@ -450,6 +450,8 @@ void OMW::Engine::setSkipMenu(bool skipMenu, bool newGame)
     mNewGame = newGame;
 }
 
+osg::ref_ptr<osgViewer::Viewer> g_viewer;
+
 void OMW::Engine::createWindow()
 {
     int screen = Settings::Manager::getInt("screen", "Video");
@@ -600,6 +602,7 @@ void OMW::Engine::createWindow()
 
     mViewer->getEventQueue()->getCurrentEventState()->setWindowRectangle(
         0, 0, graphicsWindow->getTraits()->width, graphicsWindow->getTraits()->height);
+    g_viewer = mViewer;
 }
 
 void OMW::Engine::setWindowIcon()
@@ -970,6 +973,8 @@ void OMW::Engine::go()
 
     mLuaWorker->join();
 
+    g_viewer.release();
+
     // Save user settings
     Settings::Manager::saveUser(mCfgMgr.getUserConfigPath() / "settings.cfg");
     Settings::ShaderManager::get().save();
diff --git a/components/debug/debugging.cpp b/components/debug/debugging.cpp
index 0bec74ce73b7660e377dfddcbb6e67760dc5f977..0bd0d70fcefb140fb2ad35011cd84e2c80e86821 100644
--- a/components/debug/debugging.cpp
+++ b/components/debug/debugging.cpp
@@ -34,6 +34,8 @@
 
 #include <SDL_messagebox.h>
 
+#include <osgViewer/Viewer>
+
 namespace Debug
 {
 #ifdef _WIN32
@@ -323,6 +325,8 @@ void setupLogging(const std::filesystem::path& logDir, std::string_view appName,
 #endif
 }
 
+extern osg::ref_ptr<osgViewer::Viewer> g_viewer;
+
 int wrapApplication(int (*innerApplication)(int argc, char* argv[]), int argc, char* argv[], std::string_view appName)
 {
 #if defined _WIN32
@@ -362,6 +366,7 @@ int wrapApplication(int (*innerApplication)(int argc, char* argv[]), int argc, c
     }
     catch (const std::exception& e)
     {
+    	g_viewer.release();
 #if (defined(__APPLE__) || defined(__linux) || defined(__unix) || defined(__posix))
         if (!isatty(fileno(stdin)))
 #endif
