diff --git a/extern/CMakeLists.txt b/extern/CMakeLists.txt
index 931a53de844c3dffab8d3723e3845cf2227b1859..609c668a1929d94eeba3cd98d1b06e8e3d8bf8dc 100644
--- a/extern/CMakeLists.txt
+++ b/extern/CMakeLists.txt
@@ -264,7 +264,7 @@ if (NOT OPENMW_USE_SYSTEM_ICU)
         # Wants a triple such as aarch64-linux-android, excluding a trailing
         # -clang etc.
         string(REGEX MATCH "^[^-]\+-[^-]+-[^-]+" ICU_TOOLCHAIN_NAME ${ANDROID_TOOLCHAIN_NAME})
-        set(ICU_ADDITIONAL_OPTS --host=${ICU_TOOLCHAIN_NAME}${ANDROIDVER} --with-cross-build=${CMAKE_CURRENT_BINARY_DIR}/icu-cross)
+        set(ICU_ADDITIONAL_OPTS --host=aarch64-linux-android --with-cross-build=${CMAKE_CURRENT_BINARY_DIR}/icu-cross)
         # --with-cross-build must be a build directory, not an install root, since the configure script
         # looks for a configuration file which does not get installed.
         # We need a host version of ICU so that the tools can be run when building the data library.
