diff --git a/apps/openmw/options.cpp b/apps/openmw/options.cpp
index c78dc395d9002c12186c5d8c346cca4809ee2b9b..dbf910b70f41fd5cf9678423e7743004f83b0c02 100644
--- a/apps/openmw/options.cpp
+++ b/apps/openmw/options.cpp
@@ -27,7 +27,7 @@ namespace OpenMW
             ("data", bpo::value<Files::MaybeQuotedPathContainer>()->default_value(Files::MaybeQuotedPathContainer(), "data")
                 ->multitoken()->composing(), "set data directories (later directories have higher priority)")
 
-            ("data-local", bpo::value<Files::MaybeQuotedPathContainer::value_type>()->default_value(Files::MaybeQuotedPathContainer::value_type(), ""),
+            ("data-local", bpo::value<Files::MaybeQuotedPath>()->default_value(Files::MaybeQuotedPath(), ""),
                 "set local data directory (highest priority)")
 
             ("fallback-archive", bpo::value<StringsVector>()->default_value(StringsVector(), "fallback-archive")
diff --git a/components/files/configurationmanager.cpp b/components/files/configurationmanager.cpp
index 87f4cd49430c7db6b2afe98eca95799877cf567c..80426586fed5f23b91e6fcc270f6afd3343ad392 100644
--- a/components/files/configurationmanager.cpp
+++ b/components/files/configurationmanager.cpp
@@ -168,9 +168,13 @@ void ConfigurationManager::addExtraConfigDirs(std::stack<boost::filesystem::path
 
 void ConfigurationManager::addCommonOptions(boost::program_options::options_description& description)
 {
+    Files::MaybeQuotedPath defaultUserData;
+    static_cast<boost::filesystem::path&>(defaultUserData) = boost::filesystem::path("?userdata?");
+
     description.add_options()
-        ("config", bpo::value<Files::MaybeQuotedPathContainer>()->multitoken()->composing(), "additional config directories")
-        ("user-data", bpo::value<Files::MaybeQuotedPath>(),
+        ("config", bpo::value<Files::MaybeQuotedPathContainer>()->default_value(Files::MaybeQuotedPathContainer(), "")
+            ->multitoken()->composing(), "additional config directories")
+        ("user-data", bpo::value<Files::MaybeQuotedPath>()->default_value(defaultUserData, ""),
             "set user data directory (used for saves, screenshots, etc)");
 }
 
