diff --git a/apps/openmw/mwrender/postprocessor.cpp b/apps/openmw/mwrender/postprocessor.cpp
index 08a6b4f0067..2a7f143f9d3 100644
--- a/apps/openmw/mwrender/postprocessor.cpp
+++ b/apps/openmw/mwrender/postprocessor.cpp
@@ -82,7 +82,7 @@ namespace MWRender
             ext->glDisablei = ext->glDisableIndexedEXT;
 
         if (ext->glDisablei)
-            mNormalsSupported = true;
+            mNormalsSupported = false;
         else
             Log(Debug::Error) << "'glDisablei' unsupported, pass normals will not be available to shaders.";
 
