diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
index 1a8e22ed8268de24a934f6f677e6ee61539c970c..7dc69484c6dbc9398bfc29c737f2f0d016176ae1 100644
--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -210,7 +210,6 @@ Ubuntu_GCC_Debug:
     CCACHE_SIZE: 3G
     CMAKE_BUILD_TYPE: Debug
     CMAKE_CXX_FLAGS_DEBUG: -O0
-    BUILD_SHARED_LIBS: 1
   # When CCache doesn't exist (e.g. first build on a fork), build takes more than 1h, which is the default for forks.
   timeout: 2h
 
diff --git a/CI/before_script.linux.sh b/CI/before_script.linux.sh
index ab61ed3e5963021afdfd074885cf106cde7096fd..0edd38628f81e52c6c3ba71b9eb5c3b9dc55bf1c 100755
--- a/CI/before_script.linux.sh
+++ b/CI/before_script.linux.sh
@@ -22,7 +22,7 @@ declare -a CMAKE_CONF_OPTS=(
     -DCMAKE_C_COMPILER_LAUNCHER=ccache
     -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
     -DCMAKE_INSTALL_PREFIX=install
-    -DBUILD_SHARED_LIBS="${BUILD_SHARED_LIBS:-OFF}"
+    -DBUILD_SHARED_LIBS=OFF
     -DUSE_SYSTEM_TINYXML=ON
     -DOPENMW_USE_SYSTEM_RECASTNAVIGATION=ON
     -DOPENMW_CXX_FLAGS="-Werror -Werror=implicit-fallthrough"  # flags specific to OpenMW project
diff --git a/apps/opencs/CMakeLists.txt b/apps/opencs/CMakeLists.txt
index 9bf02e10c9da33bad41e32e1831d337adeceb2a6..610c5157aac75140a026955f66b7cb39e7e54277 100644
--- a/apps/opencs/CMakeLists.txt
+++ b/apps/opencs/CMakeLists.txt
@@ -172,7 +172,7 @@ else()
     set (OPENCS_OPENMW_CFG "")
 endif(APPLE)
 
-add_library(openmw-cs-lib STATIC
+add_library(openmw-cs-lib
     ${OPENCS_SRC}
     ${OPENCS_UI_HDR}
     ${OPENCS_MOC_SRC}
diff --git a/components/CMakeLists.txt b/components/CMakeLists.txt
index dc195d8d0b989fa8879461f0d91544e8f1d1e4a3..317786e96a1df5cd57ba96bbcd21fb14d25d31d0 100644
--- a/components/CMakeLists.txt
+++ b/components/CMakeLists.txt
@@ -526,9 +526,11 @@ if (USE_QT)
     QT_WRAP_UI(ESM_UI_HDR ${ESM_UI})
 endif()
 
-if (ANDROID)
-    set_property(TARGET components PROPERTY POSTION_INDEPENDENT_CODE ON)
-endif()
+if (CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86_64" AND NOT APPLE)
+        add_definitions(-fPIC)
+    endif()
+endif ()
 
 include_directories(${BULLET_INCLUDE_DIRS} ${CMAKE_CURRENT_BINARY_DIR})
 
diff --git a/apps/openmw/CMakeLists.txt b/apps/openmw/CMakeLists.txt
index 1c11825ef..a66c4d39b 100644
--- a/apps/openmw/CMakeLists.txt
+++ b/apps/openmw/CMakeLists.txt
@@ -158,7 +158,9 @@ if(OSG_STATIC)
 endif(OSG_STATIC)
 
 if (ANDROID)
-    target_link_libraries(openmw EGL android log z)
+    find_package(OSGPlugins REQUIRED COMPONENTS ${USED_OSG_PLUGINS})
+    list(APPEND OPENSCENEGRAPH_LIBRARIES ${OSGPlugins_LIBRARIES})
+    target_link_libraries(openmw EGL android log z ${OPENSCENEGRAPH_LIBRARIES} freetype jpeg png)
 endif (ANDROID)
 
 if (USE_SYSTEM_TINYXML)
