diff --git a/components/fx/pass.cpp b/components/fx/pass.cpp
index 7a7329d755ea1de19348465eec1acf2142b00f6b..ba8044b7a5f2354a490a3d043ea6dc9446112ed6 100644
--- a/components/fx/pass.cpp
+++ b/components/fx/pass.cpp
@@ -88,7 +88,7 @@ namespace fx
 
 uniform @builtinSampler omw_SamplerLastShader;
 uniform @builtinSampler omw_SamplerLastPass;
-uniform @builtinSampler omw_SamplerDepth;
+uniform highp @builtinSampler omw_SamplerDepth;
 uniform @builtinSampler omw_SamplerNormals;
 
 uniform vec4 omw_PointLights[@pointLightCount];
