diff --git a/apps/openmw/mwgui/itemwidget.cpp b/apps/openmw/mwgui/itemwidget.cpp
index ffe854a72bfc6f651fae8bd013f6d9aac7c08754..a7e6cb8beff6eb46b387e89a0caeb95d5c7131da 100644
--- a/apps/openmw/mwgui/itemwidget.cpp
+++ b/apps/openmw/mwgui/itemwidget.cpp
@@ -25,14 +25,29 @@ namespace
         if (count == 1)
             return "";
 
+        // With small text size we can use up to 4 characters, while with large ones - only up to 3.
+        if (fontHeight > 16)
+        {
+            if (count > 999999999)
+                return MyGUI::utility::toString(count/1000000000) + "b";
+            else if (count > 9999999)
+                return ">9m";
+            else if (count > 999999)
+                return MyGUI::utility::toString(count/1000000) + "m";
+            else if (count > 9999)
+                return ">9k";
+            else if (count > 999)
+                return MyGUI::utility::toString(count/1000) + "k";
+            else
+                return MyGUI::utility::toString(count);
+        }
+
         if (count > 999999999)
             return MyGUI::utility::toString(count/1000000000) + "b";
         else if (count > 999999)
             return MyGUI::utility::toString(count/1000000) + "m";
         else if (count > 9999)
             return MyGUI::utility::toString(count/1000) + "k";
-        else if (fontHeight >= 18 && count > 999)
-            return MyGUI::utility::toString(count/1000) + "k";
         else
             return MyGUI::utility::toString(count);
     }
diff --git a/apps/openmw/mwgui/trainingwindow.cpp b/apps/openmw/mwgui/trainingwindow.cpp
index 96267cc7b241ad7f103f60d0dc1f81953c7318e5..24a4eb360736445b2fdba0cfcf6b23757850ef3a 100644
--- a/apps/openmw/mwgui/trainingwindow.cpp
+++ b/apps/openmw/mwgui/trainingwindow.cpp
@@ -98,6 +98,8 @@ namespace MWGui
         const MWWorld::Store<ESM::GameSetting> &gmst =
             MWBase::Environment::get().getWorld()->getStore().get<ESM::GameSetting>();
 
+        const int lineHeight = MWBase::Environment::get().getWindowManager()->getFontHeight() + 2;
+
         for (int i=0; i<3; ++i)
         {
             int price = static_cast<int>(pcStats.getSkill (skills[i].first).getBase() * gmst.find("iTrainingMod")->mValue.getInteger());
@@ -105,7 +107,7 @@ namespace MWGui
             price = MWBase::Environment::get().getMechanicsManager()->getBarterOffer(mPtr, price, true);
 
             MyGUI::Button* button = mTrainingOptions->createWidget<MyGUI::Button>(price <= playerGold ? "SandTextButton" : "SandTextButtonDisabled", // can't use setEnabled since that removes tooltip
-                MyGUI::IntCoord(5, 5+i*18, mTrainingOptions->getWidth()-10, 18), MyGUI::Align::Default);
+                MyGUI::IntCoord(5, 5+i*lineHeight, mTrainingOptions->getWidth()-10, lineHeight), MyGUI::Align::Default);
 
             button->setUserData(skills[i].first);
             button->eventMouseButtonClick += MyGUI::newDelegate(this, &TrainingWindow::onTrainingSelected);
diff --git a/components/fontloader/fontloader.cpp b/components/fontloader/fontloader.cpp
index e575b8cbc02f1a4d597df6e8823a99e6e55ac7e8..dbf42e810b4f4e475f38a5c006719780739fb788 100644
--- a/components/fontloader/fontloader.cpp
+++ b/components/fontloader/fontloader.cpp
@@ -155,7 +155,7 @@ namespace Gui
 
     FontLoader::FontLoader(ToUTF8::FromType encoding, const VFS::Manager* vfs, float scalingFactor)
         : mVFS(vfs)
-        , mFontHeight(std::clamp(Settings::Manager::getInt("font size", "GUI"), 12, 20))
+        , mFontHeight(std::clamp(Settings::Manager::getInt("font size", "GUI"), 12, 18))
         , mScalingFactor(scalingFactor)
     {
         if (encoding == ToUTF8::WINDOWS_1252)
@@ -176,10 +176,10 @@ namespace Gui
         loadFont("DejaVuLGCSansMono", "MonoFont"); // We need to use a TrueType monospace font to display debug texts properly.
 
         // Use our TrueType fonts as a fallback.
-        if (!MyGUI::ResourceManager::getInstance().isExist("DefaultFont") && !Misc::StringUtils::ciEqual(defaultFont, "Pelagiad"))
-            loadFont("Pelagiad", "DefaultFont");
-        if (!MyGUI::ResourceManager::getInstance().isExist("ScrollFont") && !Misc::StringUtils::ciEqual(scrollFont, "OMWAyembedt"))
-            loadFont("OMWAyembedt", "ScrollFont");
+        if (!MyGUI::ResourceManager::getInstance().isExist("DefaultFont") && !Misc::StringUtils::ciEqual(defaultFont, "MysticCards"))
+            loadFont("MysticCards", "DefaultFont");
+        if (!MyGUI::ResourceManager::getInstance().isExist("ScrollFont") && !Misc::StringUtils::ciEqual(scrollFont, "DemonicLetters"))
+            loadFont("DemonicLetters", "ScrollFont");
     }
 
     void FontLoader::loadFont(const std::string& fileName, const std::string& fontId)
@@ -553,12 +553,9 @@ namespace Gui
     std::string FontLoader::getFontForFace(const std::string& face)
     {
         const std::string lowerFace = Misc::StringUtils::lowerCase(face);
-
-        if (lowerFace == "magic cards")
-            return "DefaultFont";
         if (lowerFace == "daedric")
             return "ScrollFont";
 
-        return face;
+        return "DefaultFont";
     }
 }
diff --git a/components/widgets/fontwrapper.hpp b/components/widgets/fontwrapper.hpp
index 122cf20257e88597413df2422dc18e09d0fa7abe..3396792f348642307785263ebcf436db0d80e898 100644
--- a/components/widgets/fontwrapper.hpp
+++ b/components/widgets/fontwrapper.hpp
@@ -35,7 +35,7 @@ namespace Gui
         std::string getFontSize()
         {
             // Note: we can not use the FontLoader here, so there is a code duplication a bit.
-            static const std::string fontSize = std::to_string(std::clamp(Settings::Manager::getInt("font size", "GUI"), 12, 20));
+            static const std::string fontSize = std::to_string(std::clamp(Settings::Manager::getInt("font size", "GUI"), 12, 18));
             return fontSize;
         }
     };
diff --git a/files/data/mygui/openmw_dialogue_window.layout b/files/data/mygui/openmw_dialogue_window.layout
index d1f96121e9bdf1b86a3d69e8c4ed07ec6f34fb64..6d53d63351d860595440d38a8abd722a71775865 100644
--- a/files/data/mygui/openmw_dialogue_window.layout
+++ b/files/data/mygui/openmw_dialogue_window.layout
@@ -18,7 +18,7 @@
         <!-- The disposition bar-->
         <Widget type="ProgressBar" skin="MW_Progress_Blue" position="398 8 166 18"
             align="Right Top" name="Disposition">
-            <Widget type="TextBox" skin="ProgressText" position="0 0 166 14" name="DispositionText" align="Right VCenter">
+            <Widget type="TextBox" skin="ProgressText" position="0 0 166 12" name="DispositionText" align="Right VCenter">
                 <Property key="NeedMouse" value="false"/>
              </Widget>
         </Widget>
diff --git a/files/data/mygui/openmw_stats_window.layout b/files/data/mygui/openmw_stats_window.layout
index 77de3ac75308ee919998b101885d44dd77dca5a7..2fcede969f451a8e3492f817214e826d2fd95e16 100644
--- a/files/data/mygui/openmw_stats_window.layout
+++ b/files/data/mygui/openmw_stats_window.layout
@@ -21,7 +21,7 @@
                     </Widget>
                     <Widget type="ProgressBar" skin="MW_Progress_Red" position="74 0 130 18" name="HBar" align="Right Top">
                         <Property key="NeedMouse" value="false"/>
-                        <Widget type="TextBox" skin="ProgressText" position="0 0 130 14" name="HBarT" align="Right VCenter">
+                        <Widget type="TextBox" skin="ProgressText" position="0 0 130 12" name="HBarT" align="Right Top">
                             <Property key="NeedMouse" value="false"/>
                         </Widget>
                     </Widget>
@@ -41,7 +41,7 @@
 
                     <Widget type="ProgressBar" skin="MW_Progress_Blue" position="74 0 130 18" name="MBar" align="Right Top">
                         <Property key="NeedMouse" value="false"/>
-                        <Widget type="TextBox" skin="ProgressText" position="0 0 130 14" name="MBarT" align="Right VCenter">
+                        <Widget type="TextBox" skin="ProgressText" position="0 0 130 12" name="MBarT" align="Right VCenter">
                             <Property key="NeedMouse" value="false"/>
                         </Widget>
                     </Widget>
@@ -61,7 +61,7 @@
 
                     <Widget type="ProgressBar" skin="MW_Progress_Green" position="74 0 130 18" name="FBar" align="Right Top">
                         <Property key="NeedMouse" value="false"/>
-                        <Widget type="TextBox" skin="ProgressText" position="0 0 130 14" name="FBarT" align="Right VCenter">
+                        <Widget type="TextBox" skin="ProgressText" position="0 0 130 12" name="FBarT" align="Right VCenter">
                             <Property key="NeedMouse" value="false"/>
                         </Widget>
                     </Widget>
diff --git a/files/data/mygui/openmw_text.skin.xml b/files/data/mygui/openmw_text.skin.xml
index 5f96d0a57c838774e2c81bba7477759bb17454d8..00a69c486be8d9de27ab76beb384d9b5b260a813 100644
--- a/files/data/mygui/openmw_text.skin.xml
+++ b/files/data/mygui/openmw_text.skin.xml
@@ -149,18 +149,18 @@ color_misc=0,205,205 # ????
     <Resource type="ResourceSkin" name="MW_DynamicStat_Red" size="204 18">
         <Child type="TextBox" skin="SandText" offset="0 0 100 18" align="Left Top" name="Text"/>
         <Child type="ProgressBar" skin="MW_Progress_Red" offset="74 0 130 18" align="Right Top" name="Bar"/>
-        <Child type="TextBox" skin="ProgressText" offset="74 0 130 14" align="Right Top" name="BarText"/>
+        <Child type="TextBox" skin="ProgressText" offset="74 0 130 12" align="Right Top" name="BarText"/>
     </Resource>
 
     <Resource type="ResourceSkin" name="MW_DynamicStat_Blue" size="204 18">
         <Child type="TextBox" skin="SandText" offset="0 0 100 18" align="Left Top" name="Text"/>
         <Child type="ProgressBar" skin="MW_Progress_Blue" offset="74 0 130 18" align="Right Top" name="Bar"/>
-        <Child type="TextBox" skin="ProgressText" offset="74 0 130 14" align="Right Top" name="BarText"/>
+        <Child type="TextBox" skin="ProgressText" offset="74 0 130 12" align="Right Top" name="BarText"/>
     </Resource>
 
     <Resource type="ResourceSkin" name="MW_DynamicStat_Green" size="204 18">
         <Child type="TextBox" skin="SandText" offset="0 0 100 18" align="Left Top" name="Text"/>
         <Child type="ProgressBar" skin="MW_Progress_Green" offset="74 0 130 18" align="Right Top" name="Bar"/>
-        <Child type="TextBox" skin="ProgressText" offset="74 0 130 14" align="Right Top" name="BarText"/>
+        <Child type="TextBox" skin="ProgressText" offset="74 0 130 12" align="Right Top" name="BarText"/>
     </Resource>
 </MyGUI>
diff --git a/files/data/mygui/openmw_wait_dialog_progressbar.layout b/files/data/mygui/openmw_wait_dialog_progressbar.layout
index 8147abbdfb9bc7730abe6070f501bc6bdee3d448..720bfa4327b4a1a4d4b75f56f7aee6d0c80d5e55 100644
--- a/files/data/mygui/openmw_wait_dialog_progressbar.layout
+++ b/files/data/mygui/openmw_wait_dialog_progressbar.layout
@@ -4,7 +4,7 @@
     <Widget type="Window" skin="MW_Dialog" layer="Windows" position="0 0 219 40" align="Center" name="_Main">
 
         <Widget type="ProgressBar" skin="MW_Progress_Blue" position="5 6 199 20" name="ProgressBar">
-            <Widget type="TextBox" skin="ProgressText" position="0 0 199 16" name="ProgressText" align="Right VCenter">
+            <Widget type="TextBox" skin="ProgressText" position="0 0 199 14" name="ProgressText" align="Right VCenter">
                 <Property key="NeedMouse" value="false"/>
             </Widget>
         </Widget>
diff --git a/files/ui/advancedpage.ui b/files/ui/advancedpage.ui
index 42b46080fdbe5292cdc77997414012629877fe6d..0f895ad65bb62819e9faaf6bc8ad7ed5f3c4d697 100644
--- a/files/ui/advancedpage.ui
+++ b/files/ui/advancedpage.ui
@@ -946,7 +946,7 @@
             <number>12</number>
            </property>
            <property name="maximum">
-            <number>20</number>
+            <number>18</number>
            </property>
            <property name="singleStep">
             <number>1</number>
