diff --git a/components/shader/shadervisitor.cpp b/components/shader/shadervisitor.cpp
index fdcabedeb25d80735f833a79a11110ecf645f88c..5f7225c6fc93d344ba7e1abed5ec2a2cee7a27bf 100644
--- a/components/shader/shadervisitor.cpp
+++ b/components/shader/shadervisitor.cpp
@@ -627,10 +627,10 @@ namespace Shader
         bool simpleLighting = false;
         node.getUserValue("simpleLighting", simpleLighting);
         if (simpleLighting)
-        {
-            defineMap["forcePPL"] = "1";
             defineMap["endLight"] = "0";
-        }
+
+        if (simpleLighting || dynamic_cast<osgParticle::ParticleSystem*>(&node))
+            defineMap["forcePPL"] = "0";
 
         if (reqs.mAlphaBlend && mSupportsNormalsRT)
         {
