diff --git a/apps/openmw/mwrender/camera.cpp b/apps/openmw/mwrender/camera.cpp
index 625b6c3e97191e21e24fad5822cf85ebadfc5c5c..4dee07888e9e0ea690c80a7185fb7d2f7c050cbe 100644
--- a/apps/openmw/mwrender/camera.cpp
+++ b/apps/openmw/mwrender/camera.cpp
@@ -210,7 +210,10 @@ namespace MWRender
     void Camera::setMode(Mode newMode, bool force)
     {
         if (mMode == newMode)
+        {
+            mQueuedMode = std::nullopt;
             return;
+        }
         Mode oldMode = mMode;
         if (!force && (newMode == Mode::FirstPerson || oldMode == Mode::FirstPerson) && mAnimation && !mAnimation->upperBodyReady())
         {
diff --git a/files/data/scripts/omw/camera/camera.lua b/files/data/scripts/omw/camera/camera.lua
index 46fa732247025f72c4569dff10b4b591f367a51c..628e43b1830ced00dc44305a11146cc66529fa8d 100644
--- a/files/data/scripts/omw/camera/camera.lua
+++ b/files/data/scripts/omw/camera/camera.lua
@@ -76,6 +76,12 @@ local function updatePOV(dt)
             end
         end
         camera.setMode(primaryMode)
+        if camera.getMode() == MODE.Preview then
+            -- If Preview -> FirstPerson change is queued (because of 3rd person animation),
+            -- then first exit Preview by switching to ThirdPerson, and then queue the switch to FirstPerson.
+            camera.setMode(MODE.ThirdPerson)
+            camera.setMode(MODE.FirstPerson)
+        end
         previewTimer = 0
     end
 end
@@ -180,17 +186,15 @@ end
 local function onFrame(dt)
     if core.isWorldPaused() then return end
     local mode = camera.getMode()
-    if mode == MODE.FirstPerson or mode == MODE.ThirdPerson then
+    if (mode == MODE.FirstPerson or mode == MODE.ThirdPerson) and not camera.getQueuedMode() then
         primaryMode = mode
     end
     if mode ~= MODE.Static then
-        if not camera.getQueuedMode() or camera.getQueuedMode() == MODE.Preview then
-            if noModeControl == 0 then
-                updatePOV(dt)
-                updateVanity(dt)
-            end
-            updateStandingPreview()
+        if noModeControl == 0 then
+            updatePOV(dt)
+            updateVanity(dt)
         end
+        updateStandingPreview()
         updateCrosshair()
     end
     applyControllerZoom(dt)
