diff --git a/apps/openmw/mwstate/statemanagerimp.cpp b/apps/openmw/mwstate/statemanagerimp.cpp
index 50fd123b4f5f66ac510f2ae5b135cf6884e533b9..af46afb308acd5c800eeffc85cae2a3f3eb9a99a 100644
--- a/apps/openmw/mwstate/statemanagerimp.cpp
+++ b/apps/openmw/mwstate/statemanagerimp.cpp
@@ -17,6 +17,7 @@
 #include <components/settings/settings.hpp>
 
 #include <osg/Image>
+#include <osg/Hint>
 
 #include <osgDB/Registry>
 
@@ -322,6 +323,8 @@ void MWState::StateManager::saveGame(std::string_view description, const Slot* s
         Log(Debug::Info) << '\'' << description << "' is saved in "
                          << std::chrono::duration_cast<std::chrono::duration<float, std::milli>>(finish - start).count()
                          << "ms";
+
+        osg::Hint::GL4ES_SavePSA();
     }
     catch (const std::exception& e)
     {
