diff --git a/apps/openmw/mwstate/statemanagerimp.cpp b/apps/openmw/mwstate/statemanagerimp.cpp
index c8cd53a3310fb69efcfc37029b0420e0ef65a380..05306fdea072127bde4bd985a78ea88d0cada9a7 100644
--- a/apps/openmw/mwstate/statemanagerimp.cpp
+++ b/apps/openmw/mwstate/statemanagerimp.cpp
@@ -311,6 +311,8 @@ void MWState::StateManager::saveGame (const std::string& description, const Slot
 
         Log(Debug::Info) << '\'' << description << "' is saved in "
             << std::chrono::duration_cast<std::chrono::duration<float, std::milli>>(finish - start).count() << "ms";
+
+        osg::Hint::GL4ES_SavePSA();            
     }
     catch (const std::exception& e)
     {
