diff --git a/apps/openmw/mwrender/bulletdebugdraw.cpp b/apps/openmw/mwrender/bulletdebugdraw.cpp
index 7e979c0fec2008e93a684dde359f083dee181aab..9d7af2d001c24065c7f2890466e41a705c71741a 100644
--- a/apps/openmw/mwrender/bulletdebugdraw.cpp
+++ b/apps/openmw/mwrender/bulletdebugdraw.cpp
@@ -9,6 +9,7 @@
 
 #include <components/debug/debuglog.hpp>
 #include <components/misc/convert.hpp>
+#include <components/sceneutil/depth.hpp>
 #include <osg/PolygonMode>
 #include <osg/ShapeDrawable>
 #include <osg/StateSet>
@@ -61,6 +62,7 @@ void DebugDrawer::createGeometry()
 
         auto* stateSet = new osg::StateSet;
         stateSet->setAttributeAndModes(new osg::PolygonMode(osg::PolygonMode::FRONT_AND_BACK, osg::PolygonMode::LINE), osg::StateAttribute::ON);
+        stateSet->setAttributeAndModes(new osg::PolygonOffset(SceneUtil::AutoDepth::isReversed() ? 1.0 : -1.0, SceneUtil::AutoDepth::isReversed() ? 1.0 : -1.0));
         mShapesRoot = new osg::Group;
         mShapesRoot->setStateSet(stateSet);
         mShapesRoot->setDataVariance(osg::Object::DYNAMIC);
