diff --git a/files/shaders/water_fragment.glsl b/files/shaders/water_fragment.glsl
index eaf52c118924e365f9b9d9ac0b859b4f1392ff86..088eb1717104a7a0fcc8eaa8bee18fc67fd08c1e 100644
--- a/files/shaders/water_fragment.glsl
+++ b/files/shaders/water_fragment.glsl
@@ -210,7 +210,7 @@ uniform sampler2D normalMap;
 uniform sampler2D reflectionMap;
 #if REFRACTION
 uniform sampler2D refractionMap;
-uniform sampler2D refractionDepthMap;
+uniform highp sampler2D refractionDepthMap;
 #endif
 
 uniform float osg_SimulationTime;
@@ -368,4 +368,8 @@ void main(void)
     gl_FragData[0].xyz = mix(gl_FragData[0].xyz, gl_Fog.color.xyz, fogValue);
 
     applyShadowDebugOverlay();
+    
+    gl_FragData[0].xyz = pow(gl_FragData[0].xyz, vec3(1.0/@gamma));
+   
 }
